File: problem012/12.rb
#!/usr/bin/env ruby

#compute the number of divisors of a number
def divisors_count(num)
    sqrt = Math.sqrt(num).to_i
    exponents = []
    2.upto(sqrt).each{|i|
        p = 0
        while num % i == 0
            num = num / i
            p += 1
        end
        exponents.push p if p > 0
    }
    exponents.push 1 if num != 1
    exponents.size == 0 ? 2 : exponents.inject(1){|t, f| t *= f + 1}
end

#enumerator returning the series of triangle numbers
triangle = Enumerator.new{|t|
    i = 1
    loop {
        t.yield i * (i + 1) / 2
        i += 1
    }
}

#find the first triangle number with more than 500 divisors
triangle.each{|n|
    if divisors_count(n) > 500
        puts n
        break
    end
}

File: problem012/12.js
#!/usr/bin/env node

//lazy triangle numbers generator
var triangle = function (i) {
    var j =  i * (i + 1) / 2;
    return { "this": j, "next": function () { return triangle(i + 1); } }
}

var divisorsCount = function (num) {
    var sqrt = Math.floor(Math.sqrt(num));
    var exponents = [];
    var i = 2;
    var p;

    while (i <= sqrt) {
        p = 0;
        while (num % i == 0) {
            num /= i;
            p++;
        }
        if (p > 0) { exponents.push(p); }
        i++;
    }

    if (num != 1) { exponents.push(1); }

    return exponents.length == 0 ? 2 : exponents.reduce((prev, cur) => prev * (cur + 1), 1);
}

var t = triangle(1);

while (divisorsCount(t.this) <= 500) {
    t = t.next();
}

console.log(t.this);
